File: ta4j/src/main/java/eu/verdelhan/ta4j/BaseTradingRecord.java
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2014-2017 Marc de Verdelhan & respective authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package eu.verdelhan.ta4j;

import eu.verdelhan.ta4j.Order.OrderType;

import java.util.ArrayList;
import java.util.List;

/**
 * Base implementation of a {@link TradingRecord}.
 * <p>
 */
public class BaseTradingRecord implements TradingRecord {

	private static final long serialVersionUID = -4436851731855891220L;

	/** The recorded orders */
    private List<Order> orders = new ArrayList<Order>();
    
    /** The recorded BUY orders */
    private List<Order> buyOrders = new ArrayList<Order>();
    
    /** The recorded SELL orders */
    private List<Order> sellOrders = new ArrayList<Order>();
    
    /** The recorded entry orders */
    private List<Order> entryOrders = new ArrayList<Order>();
    
    /** The recorded exit orders */
    private List<Order> exitOrders = new ArrayList<Order>();
    
    /** The recorded trades */
    private List<Trade> trades = new ArrayList<Trade>();

    /** The entry type (BUY or SELL) in the trading session */
    private OrderType startingType;
    
    /** The current non-closed trade (there's always one) */
    private Trade currentTrade;

    /**
     * Constructor.
     */
    public BaseTradingRecord() {
        this(OrderType.BUY);
    }
    
    /**
     * Constructor.
     * @param entryOrderType the {@link OrderType order type} of entries in the trading session
     */
    public BaseTradingRecord(OrderType entryOrderType) {
        if (entryOrderType == null) {
            throw new IllegalArgumentException("Starting type must not be null");
        }
        this.startingType = entryOrderType;
        currentTrade = new Trade(entryOrderType);
    }

    /**
     * Constructor.
     * @param orders the orders to be recorded (cannot be empty)
     */
    public BaseTradingRecord(Order... orders) {
        this(orders[0].getType());
        for (Order o : orders) {
            boolean newOrderWillBeAnEntry = currentTrade.isNew();
            if (newOrderWillBeAnEntry && o.getType() != startingType) {
                // Special case for entry/exit types reversal
                // E.g.: BUY, SELL,
                //    BUY, SELL,
                //    SELL, BUY,
                //    BUY, SELL
                currentTrade = new Trade(o.getType());
            }
            Order newOrder = currentTrade.operate(o.getIndex(), o.getPrice(), o.getAmount());
            recordOrder(newOrder, newOrderWillBeAnEntry);
        }
    }
    
    @Override
    public Trade getCurrentTrade() {
        return currentTrade;
    }
    
    @Override
    public void operate(int index, Decimal price, Decimal amount) {
        if (currentTrade.isClosed()) {
            // Current trade closed, should not occur
            throw new IllegalStateException("Current trade should not be closed");
        }
        boolean newOrderWillBeAnEntry = currentTrade.isNew();
        Order newOrder = currentTrade.operate(index, price, amount);
        recordOrder(newOrder, newOrderWillBeAnEntry);
    }
    
    @Override
    public boolean enter(int index, Decimal price, Decimal amount) {
        if (currentTrade.isNew()) {
            operate(index, price, amount);
            return true;
        }
        return false;
    }
    
    @Override
    public final boolean exit(int index, Decimal price, Decimal amount) {
        if (currentTrade.isOpened()) {
            operate(index, price, amount);
            return true;
        }
        return false;
    }
    
    @Override
    public List<Trade> getTrades() {
        return trades;
    }
    
    @Override
    public Order getLastOrder() {
        if (!orders.isEmpty()) {
            return orders.get(orders.size() - 1);
        }
        return null;
    }
    
    @Override
    public Order getLastOrder(OrderType orderType) {
        if (OrderType.BUY.equals(orderType) && !buyOrders.isEmpty()) {
            return buyOrders.get(buyOrders.size() - 1);
        } else if (OrderType.SELL.equals(orderType) && !sellOrders.isEmpty()) {
            return sellOrders.get(sellOrders.size() - 1);
        }
        return null;
    }
    
    @Override
    public Order getLastEntry() {
        if (!entryOrders.isEmpty()) {
            return entryOrders.get(entryOrders.size() - 1);
        }
        return null;
    }
    
    @Override
    public Order getLastExit() {
        if (!exitOrders.isEmpty()) {
            return exitOrders.get(exitOrders.size() - 1);
        }
        return null;
    }

    /**
     * Records an order and the corresponding trade (if closed).
     * @param order the order to be recorded
     * @param isEntry true if the order is an entry, false otherwise (exit)
     */
    private void recordOrder(Order order, boolean isEntry) {
        if (order == null) {
            throw new IllegalArgumentException("Order should not be null");
        }
        
        // Storing the new order in entries/exits lists
        if (isEntry) {
            entryOrders.add(order);
        } else {
            exitOrders.add(order);
        }
        
        // Storing the new order in orders list
        orders.add(order);
        if (OrderType.BUY.equals(order.getType())) {
            // Storing the new order in buy orders list
            buyOrders.add(order);
        } else if (OrderType.SELL.equals(order.getType())) {
            // Storing the new order in sell orders list
            sellOrders.add(order);
        }

        // Storing the trade if closed
        if (currentTrade.isClosed()) {
            trades.add(currentTrade);
            currentTrade = new Trade(startingType);
        }
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/BaseTradingRecord.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core

import org.ta4j.core.Trade.TradeType
import org.ta4j.core.analysis.cost.CostModel
import org.ta4j.core.analysis.cost.ZeroCostModel
import org.ta4j.core.num.*

/**
 * Base implementation of a [TradingRecord].
 *
 */
open class BaseTradingRecord(
    entryTradeType: TradeType?,
    transactionCostModel: CostModel?,
    holdingCostModel: CostModel
) : TradingRecord {
    /**
     * The name of the trading record
     */
    override var name: String? = null

    /**
     * The recorded trades
     */
    private val trades: MutableList<Trade> = ArrayList()

    /**
     * The recorded BUY trades
     */
    private val buyTrades: MutableList<Trade> = ArrayList()

    /**
     * The recorded SELL trades
     */
    private val sellTrades: MutableList<Trade> = ArrayList()

    /**
     * The recorded entry trades
     */
    private val entryTrades: MutableList<Trade> = ArrayList()

    /**
     * The recorded exit trades
     */
    private val exitTrades: MutableList<Trade> = ArrayList()

    /**
     * The entry type (BUY or SELL) in the trading session
     */
    override val startingType: TradeType

    /**
     * The recorded positions
     */

    override val positions: MutableList<Position> = ArrayList()

    /**
     * The current non-closed position (there's always one)
     */
    private var currentPosition: Position

    /**
     * Trading cost models
     */
    private val transactionCostModel: CostModel?
    private val holdingCostModel: CostModel

    /**
     * Constructor.
     *
     * @param name the name of the tradingRecord
     */
    constructor(name: String?) : this(TradeType.BUY) {
        this.name = name
    }

    /**
     * Constructor.
     *
     * @param name           the name of the trading record
     * @param entryTradeType the [trade type][TradeType] of entries in the
     * trading session
     */
    constructor(name: String?, tradeType: TradeType?) : this(tradeType, ZeroCostModel(), ZeroCostModel()) {
        this.name = name
    }
    /**
     * Constructor.
     *
     * @param entryTradeType the [trade type][TradeType] of entries in the
     * trading session
     */
    /**
     * Constructor.
     */
    @JvmOverloads
    constructor(tradeType: TradeType? = TradeType.BUY) : this(tradeType, ZeroCostModel(), ZeroCostModel()) {
    }

    /**
     * Constructor.
     *
     * @param entryTradeType       the [trade type][TradeType] of entries in
     * the trading session
     * @param transactionCostModel the cost model for transactions of the asset
     * @param holdingCostModel     the cost model for holding asset (e.g. borrowing)
     */
    init {
        requireNotNull(entryTradeType) { "Starting type must not be null" }
        startingType = entryTradeType
        this.transactionCostModel = transactionCostModel
        this.holdingCostModel = holdingCostModel
        currentPosition = Position(entryTradeType, transactionCostModel, holdingCostModel)
    }

    /**
     * Constructor.
     *
     * @param trades the trades to be recorded (cannot be empty)
     */
    constructor(vararg trades: Trade) : this(ZeroCostModel(), ZeroCostModel(), *trades) {}

    /**
     * Constructor.
     *
     * @param transactionCostModel the cost model for transactions of the asset
     * @param holdingCostModel     the cost model for holding asset (e.g. borrowing)
     * @param trades               the trades to be recorded (cannot be empty)
     */
    constructor(
        transactionCostModel: CostModel?,
        holdingCostModel: CostModel,
        vararg trades: Trade
    ) : this(trades[0].type, transactionCostModel, holdingCostModel) {
        for (o in trades) {
            val newTradeWillBeAnEntry = currentPosition.isNew
            if (newTradeWillBeAnEntry && o.type !== startingType) {
                // Special case for entry/exit types reversal
                // E.g.: BUY, SELL,
                // BUY, SELL,
                // SELL, BUY,
                // BUY, SELL
                currentPosition = Position(o.type, transactionCostModel, holdingCostModel)
            }
            val newTrade = currentPosition.operate(o.index, o.pricePerAsset, o.amount)
            recordTrade(newTrade, newTradeWillBeAnEntry)
        }
    }



    override fun getCurrentPosition(): Position {
        return currentPosition
    }

    override fun operate(index: Int, price: Num?, amount: Num?) {
        check(!currentPosition.isClosed) {
            // Current position closed, should not occur
            "Current position should not be closed"
        }
        val newTradeWillBeAnEntry = currentPosition.isNew
        val newTrade = currentPosition.operate(index, price, amount!!)
        recordTrade(newTrade, newTradeWillBeAnEntry)
    }

    override fun enter(index: Int, price: Num?, amount: Num?): Boolean {
        if (currentPosition.isNew) {
            operate(index, price, amount)
            return true
        }
        return false
    }

    override fun exit(index: Int, price: Num?, amount: Num?): Boolean {
        if (currentPosition.isOpened) {
            operate(index, price, amount)
            return true
        }
        return false
    }

    override fun getLastTrade(): Trade? {
        return if (!trades.isEmpty()) {
            trades[trades.size - 1]
        } else null
    }

    override fun getLastTrade(tradeType: TradeType): Trade? {
        if (TradeType.BUY === tradeType && !buyTrades.isEmpty()) {
            return buyTrades[buyTrades.size - 1]
        } else if (TradeType.SELL === tradeType && !sellTrades.isEmpty()) {
            return sellTrades[sellTrades.size - 1]
        }
        return null
    }

    override fun getLastEntry(): Trade? {
        return if (!entryTrades.isEmpty()) {
            entryTrades[entryTrades.size - 1]
        } else null
    }

    override fun getLastExit(): Trade? {
        return if (!exitTrades.isEmpty()) {
            exitTrades[exitTrades.size - 1]
        } else null
    }

    /**
     * Records a trade and the corresponding position (if closed).
     *
     * @param trade   the trade to be recorded
     * @param isEntry true if the trade is an entry, false otherwise (exit)
     */
    private fun recordTrade(trade: Trade?, isEntry: Boolean) {
        requireNotNull(trade) { "Trade should not be null" }

        // Storing the new trade in entries/exits lists
        if (isEntry) {
            entryTrades.add(trade)
        } else {
            exitTrades.add(trade)
        }

        // Storing the new trade in trades list
        trades.add(trade)
        if (TradeType.BUY === trade.type) {
            // Storing the new trade in buy trades list
            buyTrades.add(trade)
        } else if (TradeType.SELL === trade.type) {
            // Storing the new trade in sell trades list
            sellTrades.add(trade)
        }

        // Storing the position if closed
        if (currentPosition.isClosed) {
            positions.add(currentPosition)
            currentPosition = Position(startingType, transactionCostModel, holdingCostModel)
        }
    }

    override fun toString(): String {
        val sb = StringBuilder()
        sb.append("BaseTradingRecord: " + if (name == null) "" else name)
        sb.append(System.lineSeparator())
        for (trade in trades) {
            sb.append(trade.toString()).append(System.lineSeparator())
        }
        return sb.toString()
    }

    companion object {
        private const val serialVersionUID = -4436851731855891220L
    }
}